storage/db.py

```
@@ -1035,6 +1167,156 @@ async def upsert_data_health_snapshots(
+        cfv_params = serialize_batch(CONSENSUS_FEATURE_SPEC, consensus_feature_vectors)
+        ps_params = serialize_batch(PERCENTILE_SNAPSHOT_SPEC, percentile_snapshots)
+        dhs_params = serialize_batch(DATA_HEALTH_SNAPSHOT_SPEC, data_health_snapshots)
+
+        if not ps_params and not (
+            isinstance(percentile_snapshots_absent_reason, str)
+            and percentile_snapshots_absent_reason.strip()
+        ):
+            raise ValueError(
+                "publish_stage2_correction: percentile_snapshots is empty and no "
+                "percentile_snapshots_absent_reason was given -- publication must not "
+                "silently omit the percentile family (§13)")
```

Ruff (0.16.1)

[warning] Avoid specifying long messages outside the exception class

(TRY003)